### PR TITLE
address gosec issues

### PR DESCRIPTION
**Pull request by wenfeng-nextmv**
Address issues found by gosec and remove all rule exclusions from security scan action.

### CODE REVIEW

File: nextroute/common/utils.go [hunk 1]
```diff
@@ -262,20 +269,43 @@ func RandomElementIndices[T any](
...
 	if source == nil {
+		// using unsafe math/rand is just a unlikely fallback, we can ignore the gosec issue here
+		// G404 (CWE-338): Use of weak random number generator (math/rand instead of crypto/rand)
+		/* #nosec */
 		source = rand.New(rand.NewSource(time.Now().UnixNano()))
 	}
 
 	for {
-		index := source.Intn(size)
+		index := RandIntnWithFallback(source, size)
```
Comment (larsbeck) [inline nextroute/common/utils.go]: We are using `RandomIndex` as part of our algorithm. Pseudo-random number generators are fine in this context. We do not worry about them not being cryptographically safe. What is more important is that they are fast. Unless there is something I am missing, I would like us to keep using `rand.Rand`.

File: route/osrm/client.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func (c *client) get(uri string) (data []byte, err error) {
 	var key string
 
 	if c.useCache {
-		key = fmt.Sprintf("%x", sha1.Sum([]byte(uri)))
+		key = fmt.Sprintf("%x", sha256.Sum256([]byte(uri)))
```
Comment (larsbeck) [inline route/osrm/client.go]: Same here. Speed matters the most in this case and this is not a security relevant operation. We are creating a key for a cache. With that said `sha1` is probably also not the best (fastest) choice, but I do not see the need to use `sha256`.

**Review by larsbeck — DISMISSED** [review]

**Comment by larsbeck** [comment]
I think we should keep the `gosec` exceptions in here and not change the code as these are algorithmic internals that do rely on speed, but not on cryptographic safety. Maybe I am missing how an attacker might use the fact that we are using non-cryptographically safe random stream to make random decisions or break ties in the algorithm. If so, could you give an example of how that would be done?

File: nextroute/common/utils.go (hunk 1 above)
Comment (wenfeng-nextmv) [reply nextroute/common/utils.go]: Revert to use math/rand and add #nosec after discussion.

File: route/osrm/client.go (hunk 2 above)
Comment (wenfeng-nextmv) [reply route/osrm/client.go]: Revert to sha1 and add #nosec after discussion.

**Review by davidrijsman — APPROVED** [review]

File: templates/measure-matrix/main.go [hunk 3]
```diff
@@ -109,6 +109,13 @@ func main() {
 
 	// Write the output.
 	file, _ := json.MarshalIndent(out, "", " ")
+
+	// [gosec issue] G306 (CWE-276): Expect WriteFile permissions to be 0600 or less
+	// macOS default file permission is 0644 -rw-r--r--
+	// Linux default file permission is 0660 -rw-rw-r--
+	// This file we created contains no secret or credential. 0644 permission should not have any security risk.
+	// If a user adds sensitive content in this file later, it is the user's responsibility to protect it.
+	/* #nosec */
```
Comment (merschformann) [inline templates/measure-matrix/main.go]: This file is part of a template that will be user facing (generate via: `nextmv sdk init -t measure-matrix`). I think we should avoid having such internal comments in there. Maybe we can add the template dir to the exclusions?

File: .github/workflows/gosec.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ jobs:
         with:
           # For now, exclude some rules that we currently have issues with, so that PR merges won't be blocked.
           # After we address those issues, we shall remove the excluded rules.
-          args: "-no-fail -exclude=G306,G401,G404,G505 -fmt csv -out results.csv -stdout -verbose=text -exclude-generated -exclude-dir=template.go -exclude-dir=sample_code -exclude-dir=node_modules/aws-cdk -exclude-dir=cdk.out ./..."
+          args: "-no-fail -fmt csv -out results.csv -stdout -verbose=text -exclude-generated -exclude-dir=template.go -exclude-dir=sample_code -exclude-dir=node_modules/aws-cdk -exclude-dir=cdk.out ./..."
```
Comment (merschformann) [inline .github/workflows/gosec.yml]: I think there are some copy / paste exclusions here that are not relevant for the sdk repo, right? :thinking: 
I.e., `sample_code`, `node_modules/aws-cdk` & `cdk.out`?

File: route/osrm/client.go [hunk 5]
```diff
@@ -128,6 +131,11 @@ func (c *client) get(uri string) (data []byte, err error) {
 	var key string
 
 	if c.useCache {
+		// sha1 is used to shorten the key for faster cache lookup than directly using the lenthy uri as key.
+		// The chance of hash colision is extremely low for sha1.
+		// The cache is local to the user, which won't become a secrity threat even when key colides.
```
Comment (merschformann) [inline route/osrm/client.go]:
```suggestion
		// The cache is local to the user, which won't become a security threat even when key collides.
```
Just a small typo. :relaxed:

**Review by merschformann — COMMENTED** [review]
Thanks for the work! :hugs: 
Just a small remark about the comment in the `templates/...` file. Maybe we should exclude that whole directory? :thinking:

File: .github/workflows/gosec.yml (hunk 4 above)
Comment (wenfeng-nextmv) [reply .github/workflows/gosec.yml]: You are right. These are for cloud/ and not needed here.

File: route/osrm/client.go (hunk 5 above)
Comment (wenfeng-nextmv) [reply route/osrm/client.go]: Fixed. Good eyes!

File: templates/measure-matrix/main.go (hunk 3 above)
Comment (wenfeng-nextmv) [reply templates/measure-matrix/main.go]: I am removing most of the comment but leaving `#nosec`. For now I am a little hesitated to exclude the entire `templates/` dir from gosec scan. After all, we are responsible to provide good, secure template code samples for users to get started safely.

**Review by cowanwalls — APPROVED** [review]

**Pull request merged**